src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/RuntimeTypeMetadataUpdateHandler.cs

```diff
@@ -10,16 +12,60 @@ namespace System.Reflection.Metadata
     /// <summary>Metadata update handler used to clear a Type's reflection cache in response to a metadata update notification.</summary>
     internal static class RuntimeTypeMetadataUpdateHandler
     {
-        public static void BeforeUpdate(Type? type)
+        /// <summary>Clear type caches in response to an update notification.</summary>
+        /// <param name="types">The specific types to be cleared, or null to clear everything.</param>
+        public static void BeforeUpdate(Type[]? types)
         {
-            if (type is RuntimeType rt)
+            if (RequiresClearingAllTypes(types))
             {
-                rt.ClearCache();
+                // Clear the caches on all loaded types.
+                foreach (AssemblyLoadContext alc in AssemblyLoadContext.All)
+                {
+                    foreach (Assembly assembly in alc.Assemblies)
+                    {
+                        foreach (Type type in assembly.GetTypes())
+                        {
+                            ClearCache(type);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                // Clear the caches on just the specified types.
+                foreach (Type type in types)
+                {
+                    ClearCache(type);
+                }
+            }
+        }
+
+        /// <summary>Clears the cache on the specified type.</summary>
+        private static void ClearCache(Type type) => (type as RuntimeType)?.ClearCache();
+
+        /// <summary>Determines whether we need to clear the caches on all types or just the ones specified.</summary>
+        /// <param name="types">The types to examine.</param>
+        /// <returns>true if we need to clear all types; false if we can clear just the ones specified.</returns>
+        private static bool RequiresClearingAllTypes([NotNullWhen(false)] Type[]? types)
+        {
+            // If we were handed null, assume we need to clear everything.
+            if (types is null)
+            {
+                return true;
+            }
+
+            // If any type isn't sealed, we may need to clear a derived type,
+            // so clear everything.
+            foreach (Type type in types)
+            {
+                if (!type.IsSealed)
+                {
+                    return true;
```

Comment:
For webapps, most of the types that users operate on are not sealed - mostly because our project templates never seal types, and often intentionally because we want users to be able to derive. 

If the result usually remains the same regardless of whether dotnet-watch / VS calculates the list of types that changed, should we not bother calculating the list of changed types?

Reply:
Might other handlers want to know the types being updated?

Reply:
Hypothetically yes - perhaps the app might want to do different things depending on what changed. E.g. if you're hosting a a Blazor WebView in a Winforms / WPF app you might want to trigger a Blazor re-render but not a WinForms re-render if a component changed; but we haven't gotten far enough.